trieste/models/config.py

```diff
@@ -21,93 +21,135 @@
+    This is a registry of all supported models with their corresponding wrappers, and model
+    optimizers with their corresponding wrapppers.
+
+    A single entry per model and optimizer is maintained, if same model is registered again it will
+    overwrite the previous entry. Registry is primarily used by :class:`ModelConfig` and
+    :func:`create_model` to facilitate building models by specifying a simple dictionary of model
+    and optimizer arguments.
 
     Note that you do not need to register your custom model if you will provide an instance of
     :class:`~trieste.models.TrainableProbabilisticModel` directly to the
     :class:`~trieste.BayesianOptimizer`. Registering is required if you intend to build your custom
-    model as a dictionary of arguments for the interface and the optimizer, or as a
+    model as a dictionary of arguments for the wrapper and the optimizer, or as a
     :class:`ModelConfig`.
     """
 
-    _REGISTRY: Dict[Type[Any], Tuple[Type[TrainableProbabilisticModel], Type[Optimizer]]] = dict()
+    _MODEL_TO_WRAPPER: Dict[Type[Any], Type[TrainableProbabilisticModel]] = dict()
+    _OPTIMIZER_TO_WRAPPER: Dict[Type[Any], Type[Optimizer]] = dict()
 
     @classmethod
-    def get_interface(cls, model_type: Type[Any]) -> Type[TrainableProbabilisticModel]:
+    def get_model_wrapper(cls, model_type: Type[Any]) -> Type[TrainableProbabilisticModel]:
         """
-        Get a Trieste model interface for a given model type.
+        Get a Trieste model wrapper for a given model type.
 
         :param model_type: The model type.
-        :return: The interface which builds a model.
+        :return: The wrapper which builds a model.
         """
         try:
-            return cls._REGISTRY[model_type][0]
+            return cls._MODEL_TO_WRAPPER[model_type]
         except KeyError as e:
             raise ValueError(
                 f"Model {type(model_type)} currently not supported. "
                 f"We only support the following model types: "
-                f"{list(k.__name__ for k in cls._REGISTRY.keys())}. "
+                f"{list(k.__name__ for k in cls._MODEL_TO_WRAPPER.keys())}. "
                 f"You can register more via ModelRegistry.register_model."
             ) from e
 
     @classmethod
-    def get_optimizer(cls, model_type: Type[Any]) -> Type[Optimizer]:
+    def get_optimizer_wrapper(cls, optimizer_type: Type[Any]) -> Type[Optimizer]:
         """
-        Get a Trieste model optimizer for a given model type.
+        Get a Trieste model optimizer wrapper for a given optimizer type.
 
-        :param model_type: The model type.
-        :return: The optimizer to be used with the model type.
+        :param optimizer_type: The model type.
+        :return: The optimizer wrapper to be used with the optimizer type.
         """
         try:
-            return cls._REGISTRY[model_type][1]
+            return cls._OPTIMIZER_TO_WRAPPER[optimizer_type]
```

Comment:
(this comment applies equally well to the model registry)

Would it be better instead if we check if type is a _subclass_ of any of the registered types? That way we wouldn't have to explicitly register all the subclasses of tf.optimizers.Optimizer, and the registry wouldn't fail for trivial model subclasses (e.g. the ones we currently recommend people write to add new tensorboard logs).

In fact the best might be to loop over the `type.__mro__` types in order, and use the first match we find. That way we can define a more specific wrapper for a subclass if we want to, but don't have to if we don't.

Reply:
yes, that would be an improvement, thanks for the suggestion!
what would be the best way to check if type is a subclass? using `issubclass` function?

not sure I understand the second part, you mean to do this when choosing what to return?

Reply:
the `__mro__` property lists al the superclasses of a type in order (so starting with the type itself), so you can loop over them and check if any of them are registered.

```python
for optimizer_type in optimizer_type.__mro__:
    if optimizer_type in cls._OPTIMIZER_TO_WRAPPER:
        return cs._OPTIMIZER_TO_WRAPPER[optimizer_type]
else:
   raise ...
````

Reply:
changed now following your proposal!

Reply:
on some further reflection, I'm not sure configs can deal with this "the registry wouldn't fail for trivial model subclasses (e.g. the ones we currently recommend people write to add new tensorboard logs)" - with this modification we are checking GPR, SVGP and their superclasses, we never check for subclasses of our model wrappers, right?

that might be a bit of a gap in our configs (both old ones and this new version), if a same model can be used with multiple wrappers, configs would not know which one to use - so as soon as you want to use a non-default wrapper, you cannot use configs, or you have to overwrite the model-wrapper mapping in the registry

Reply:
With this change we are checking all the supertypes of the optimizer types against the registered types. So if we pass in a GPR subclass then we would also check it against GPR.

Reply:
right, but thats GPR as a GPflow object, this wouldn't work for subclasses of say GaussianProcessRegression wrapper, and these are the subclasses we care about, we would rarely create subclasses of GPR I think

Reply:
yes you're right. if people want to a ModelConfig with a wrapper subclass they would have to register it. though more likely they'll just construct the model wrapper themselves.